File: java/src/main/java/com/languageweaver/sdk/samples/feedback/FeedbackService.java
package com.languageweaver.sdk.samples.feedback;

import com.languageweaver.sdk.common.LanguageWeaverClient;
import com.languageweaver.sdk.common.SdkFactory;
import com.languageweaver.sdk.common.configurations.ClientConfiguration;
import com.languageweaver.sdk.common.constants.ApprovalStatus;
import com.languageweaver.sdk.common.constants.FeedbackComment;
import com.languageweaver.sdk.feedback.common.request.*;
import com.languageweaver.sdk.feedback.common.result.FeedbackListResult;
import com.languageweaver.sdk.feedback.common.result.FeedbackResult;

import java.util.List;

public class FeedbackService {

    public static FeedbackResult createFeedback() throws Exception {
        try (LanguageWeaverClient lwClient = new SdkFactory().getLanguageWeaverClient(new ClientConfiguration())) {
            CreateFeedbackRequest createFeedbackRequest = new CreateFeedbackRequest()
                    .setSourceLanguageId("eng")
                    .setTargetLanguageId("fra")
                    .setModel("generic")
                    .setSourceText("sourceText")
                    .setTargetText("targetText")
                    .setSuggestedTranslation("suggestedTranslation")
                    .addComment(FeedbackComment.SPELLING);

            return lwClient.createFeedback(createFeedbackRequest);
        }
    }

    public static void updateFeedback() throws Exception {
        try (LanguageWeaverClient lwClient = new SdkFactory().getLanguageWeaverClient(new ClientConfiguration())) {
            UpdateFeedbackRequest updateFeedbackRequest = new UpdateFeedbackRequest()
                    .setFeedbackId("feedbackId")
                    .setSuggestedTranslation("new suggested translation")
                    .addComment(FeedbackComment.CAPITALIZATION_PUNCTUATION);

            FeedbackResult feedback = lwClient.updateFeedback(updateFeedbackRequest);
        }
    }

    public static void updateFeedbackStatus() throws Exception {
        try (LanguageWeaverClient lwClient = new SdkFactory().getLanguageWeaverClient(new ClientConfiguration())) {
            UpdateFeedbackApprovalRequest updateApprovalRequest = new UpdateFeedbackApprovalRequest()
                    .setFeedbackId("feedbackId")
                    .setApprovalStatus(ApprovalStatus.REJECTED);

            FeedbackResult feedback = lwClient.updateFeedbackApproval(updateApprovalRequest);
            System.out.println(feedback.getApprovalState());
        }
    }

    public static void deleteFeedback() throws Exception {
        try (LanguageWeaverClient lwClient = new SdkFactory().getLanguageWeaverClient(new ClientConfiguration())) {
            DeleteFeedbackRequest deleteFeedbackRequest = new DeleteFeedbackRequest()
                    .setFeedbackId("feedbackId");

            lwClient.deleteFeedback(deleteFeedbackRequest);
        }
    }

    public static void retrieveFeedback() throws Exception {
        try (LanguageWeaverClient lwClient = new SdkFactory().getLanguageWeaverClient(new ClientConfiguration())) {
            int pageNumber = 1;
            int pageSize = 50;
            FeedbackListResult feedbackListResult;
            do {
                feedbackListResult = lwClient.getFeedback(pageNumber, pageSize);
                // process feedback
                List<FeedbackResult> feedbackList = feedbackListResult.getFeedbackList();
                feedbackList.stream()
                        .limit(3)
                        .forEach(feedbackResult -> System.out.println(feedbackResult.getFeedbackId()));
                pageNumber++;
            } while (feedbackListResult.getFeedbackList().size() == pageSize);
        }
    }

    public static void filterFeedback() throws Exception {
        try (LanguageWeaverClient lwClient = new SdkFactory().getLanguageWeaverClient(new ClientConfiguration())) {
            FilterFeedbackRequest filterFeedbackRequest = new FilterFeedbackRequest()
                    .setApprovalStatus(ApprovalStatus.APPROVED)
                    .setUser("userId")
                    .setStartDate("10/10/2021")
                    .setEndDate("10/10/2022");

            int pageNumber = 1;
            FeedbackListResult filteredFeedback;
            do {
                filteredFeedback = lwClient.getFeedback(filterFeedbackRequest, pageNumber);
                // process feedback
                List<FeedbackResult> feedbackList = filteredFeedback.getFeedbackList();
                feedbackList.stream()
                        .limit(3)
                        .forEach(feedbackResult -> System.out.println(feedbackResult.getFeedbackId()));
                pageNumber++;
            } while (filteredFeedback.getFeedbackList().size() > 0);
        }
    }


    public static void main(String[] args) throws Exception {
        try (LanguageWeaverClient lwClient = new SdkFactory().getLanguageWeaverClient(new ClientConfiguration())) {
            FeedbackResult feedback;

            // CREATE FEEDBACK
            feedback = createFeedback();
            System.out.println(feedback.getFeedbackId());

            // UPDATE FEEDBACK
            UpdateFeedbackRequest updateFeedbackRequest = feedback.toUpdateRequest()
                    .setSuggestedTranslation("new suggested translation")
                    .clearComments().addComment(FeedbackComment.CAPITALIZATION_PUNCTUATION);
            feedback = lwClient.updateFeedback(updateFeedbackRequest);

            // UPDATE FEEDBACK APPROVAL
            UpdateFeedbackApprovalRequest updateApprovalRequest = feedback.toUpdateApprovalRequest(ApprovalStatus.REJECTED);
            feedback = lwClient.updateFeedbackApproval(updateApprovalRequest);
            System.out.println(feedback.getApprovalState());

            // DELETE FEEDBACK
            DeleteFeedbackRequest deleteFeedbackRequest = feedback.toDeleteRequest();
            lwClient.deleteFeedback(deleteFeedbackRequest);
        }
    }
}

File: dotnet/Samples/Feedback/FeedbackService.cs
using LanguageWeaver.Sdk.Common;
using LanguageWeaver.Sdk.Configurations;
using LanguageWeaver.Sdk.Constants;
using LanguageWeaver.Sdk.Feedback.Common.Request;
using LanguageWeaver.Sdk.Feedback.Common.Result;

namespace LanguageWeaver.Sdk.Samples.Feedback;

public class FeedbackService
{
    public static void Main()
    {
        using var lwClient = new SdkFactory().GetLanguageWeaverClient(new ClientConfiguration());

        FeedbackResult feedback;

        //CREATE FEEDBACK
        feedback = CreateFeedback();
        Console.WriteLine(feedback.FeedbackId);

        //UPDATE FEEDBACK
        UpdateFeedbackRequest updateFeedbackRequest = feedback.ToUpdateRequest();
        updateFeedbackRequest.SuggestedTranslation = "new suggested translation";
        updateFeedbackRequest.ClearComments().AddComment(FeedbackComment.CapitalizationPunctuation);
        feedback = lwClient.UpdateFeedback(updateFeedbackRequest);

        //UPDATE FEEDBACK APPROVAL
        UpdateFeedbackApprovalRequest updateFeedbackApprovalRequest =
            feedback.ToUpdateApprovalRequest(ApprovalStatus.Rejected);
        feedback = lwClient.UpdateFeedbackApproval(updateFeedbackApprovalRequest);
        Console.WriteLine(feedback.ApprovalState);

        //DELETE FEEDBACK
        DeleteFeedbackRequest deleteFeedbackRequest = feedback.ToDeleteRequest();
        lwClient.DeleteFeedback(deleteFeedbackRequest);
    }

    public static FeedbackResult CreateFeedback()
    {
        using var lwClient = new SdkFactory().GetLanguageWeaverClient(new ClientConfiguration());

        CreateFeedbackRequest createFeedbackRequest = new CreateFeedbackRequest()
        {
            SourceLanguageId = "eng",
            TargetLanguageId = "fra",
            Model = "generic",
            SourceText = "sourceText",
            TargetText = "targetText",
            SuggestedTranslation = "suggestedTranslation",
        };
        createFeedbackRequest.AddComment(FeedbackComment.Spelling);

        FeedbackResult feedback = lwClient.CreateFeedback(createFeedbackRequest);

        return feedback;
    }

    public static void UpdateFeedback()
    {
        using var lwClient = new SdkFactory().GetLanguageWeaverClient(new ClientConfiguration());

        UpdateFeedbackRequest updateFeedbackRequest = new UpdateFeedbackRequest()
        {
            FeedbackId = "feedbackId",
            SuggestedTranslation = "new suggested translation"
        };

        updateFeedbackRequest.AddComment(FeedbackComment.CapitalizationPunctuation);

        FeedbackResult feedback = lwClient.UpdateFeedback(updateFeedbackRequest);
    }

    public static void UpdateFeedbackStatus()
    {
        using var lwClient = new SdkFactory().GetLanguageWeaverClient(new ClientConfiguration());

        UpdateFeedbackApprovalRequest updateFeedbackApprovalRequest = new UpdateFeedbackApprovalRequest()
        {
            FeedbackId = "feedbackId",
            ApprovalStatus = ApprovalStatus.Approved
        };

        FeedbackResult feedback = lwClient.UpdateFeedbackApproval(updateFeedbackApprovalRequest);
        Console.WriteLine(feedback.ApprovalState);
    }

    public static void DeleteFeedback()
    {
        using var lwClient = new SdkFactory().GetLanguageWeaverClient(new ClientConfiguration());

        DeleteFeedbackRequest deleteFeedbackRequest = new DeleteFeedbackRequest()
        {
            FeedbackId = "feedbackId"
        };

        lwClient.DeleteFeedback(deleteFeedbackRequest);
    }

    public static void RetrieveFeedback()
    {
        using var lwClient = new SdkFactory().GetLanguageWeaverClient(new ClientConfiguration());

        int pageNumber = 1;
        int pageSize = 50;
        FeedbackListResult feedbackListResult;
        do
        {
            feedbackListResult = lwClient.GetFeedback(pageNumber, pageSize);
            //process feedback
            List<FeedbackResult> feedbackList = feedbackListResult.FeedbackList;
            feedbackList.Take(3)
                .ToList()
                .ForEach(feedbackResult => Console.WriteLine(feedbackResult.FeedbackId));
            pageNumber++;
        } while (feedbackListResult.FeedbackList.Count == pageSize);
    }

    public static void FilterFeedback()
    {
        using var lwClient = new SdkFactory().GetLanguageWeaverClient(new ClientConfiguration());

        FilterFeedbackRequest filterFeedbackRequest = new FilterFeedbackRequest()
        {
            ApprovalStatus = ApprovalStatus.Approved,
            User = "userId",
            StartDate = "10/10/2021",
            EndDate = "10/10/2022"
        };

        int pageNumber = 1;
        FeedbackListResult filteredFeedback;
        do
        {
            filteredFeedback = lwClient.GetFeedback(filterFeedbackRequest, pageNumber);
            //process feedback
            List<FeedbackResult> feedbackList = filteredFeedback.FeedbackList;
            feedbackList.Take(3)
                .ToList()
                .ForEach(feedbackResult => Console.WriteLine(feedbackResult.FeedbackId));
            pageNumber++;
        } while (filteredFeedback.FeedbackList.Count > 0);
    }
}